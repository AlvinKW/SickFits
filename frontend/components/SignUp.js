import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';

import Form from './styles/Form';
import Error from './ErrorMessage';

const SIGNUP_MUTATION = gql`
	mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
		signUp(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

class SignUp extends Component {
	state = {
		name: '',
		email: '',
		password: '',
	};

	saveToState = event => {
		this.setState({ [event.target.name]: event.target.value });
	};

	render() {
		return (
			<Mutation mutation={SIGNUP_MUTATION} variables={this.state}>
				{(signUp, { loading, error }) => (
					<Form method="post" onSubmit={async event => {
						event.preventDefault();
						await signUp();
						this.setState({ name: '', email: '', password: '' });
					}}>
						<fieldset disabled={loading} aria-busy={loading}>
							<h2>Sign Up For An Account</h2>
							<Error error={error} />
							<label htmlFor="name">
								Name
								<input
									type="text"
									name="name"
									placeholder="Name"
									value={this.state.name}
									onChange={this.saveToState}
								/>
							</label>
							<label htmlFor="email">
								Email
								<input
									type="email"
									name="email"
									placeholder="E-Mail"
									value={this.state.email}
									onChange={this.saveToState}
								/>
							</label>
							<label htmlFor="password">
								Password
								<input
									type="password"
									name="password"
									placeholder="Password"
									value={this.state.password}
									onChange={this.saveToState}
								/>
							</label>
							<button type="submit">Sign Up</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		);
	}
}

export default SignUp;
