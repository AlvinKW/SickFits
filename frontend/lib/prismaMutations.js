import gql from 'graphql-tag';

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($id: ID!) {
		addToCart(id: $id) {
			id
			quantity
		}
	}
`;

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION (
		$title: String!
		$description: String!
		$image: String
		$largeImage: String
		$price: Int!
	) {
		createItem(
			title: $title
			description: $description
			image: $image
			largeImage: $largeImage
			price: $price
		) {
			id
		}
	}
`;

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token: String!) {
		createOrder(token: $token) {
			id
			total
			charge
			items {
				id
				title
			}
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
		removeFromCart(id: $id) {
			id
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const RESET_PASSWORD_MUTATION = gql`
	mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
		resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
			id
			name
			email
		}
	}
`;

export const SIGN_IN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signIn(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGN_OUT_MUTATION = gql`
	mutation SIGN_OUT_MUTATION {
		signOut {
			message
		}
	}
`;

export const SIGN_UP_MUTATION = gql`
	mutation SIGNUP_MUTATION($name: String!, $email: String!, $password: String!) {
		signUp(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const TOGGLE_CART_MUTATION = gql`
	mutation TOGGLE_CART_MUTATION {
		toggleCart @client
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation UPDATE_ITEM_MUTATION (
		$id: ID!
		$title: String
		$description: String
		$price: Int
	) {
		updateItem(
			id: $id
			title: $title
			description: $description
			price: $price
		) {
			id
			title
			description
			price
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userID: ID!) {
		updatePermissions(permissions: $permissions, userID: $userID) {
			id
			name
			email
			permissions
		}
	}
`;
