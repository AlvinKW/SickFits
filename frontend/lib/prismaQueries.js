import gql from 'graphql-tag';

import { perPage } from '../config';

export const ALL_ITEMS_QUERY = gql`
	query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
		items(orderBy: createdAt_DESC, skip: $skip, first: $first) {
			id
			title
			description
			image
			largeImage
			price
		}
	}
`;

export const ALL_USERS_QUERY = gql`
	query ALL_USERS_QUERY {
		users {
			id
			name
			email
			permissions
		}
	}
`;

export const CURRENT_USER_QUERY = gql`
	query CURRENT_USER_QUERY {
		me {
			id
			name
			email
			permissions
			orders {
				id
			}
			cart {
				id
				quantity
				item {
					id
					title
					description
					image
					price
				}
			}
		}
	}
`;

export const LOCAL_STATE_QUERY = gql`
	query LOCAL_STATE_QUERY {
		cartOpen @client
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

export const SEARCH_ITEMS_QUERY = gql`
	query SEARCH_ITEMS_QUERY($searchTerm: String!) {
		items(where: { OR: [{ title_contains: $searchTerm }, { description_contains: $searchTerm }] }) {
			id
			title
			image
		}
	}
`;

export const SINGLE_ITEM_QUERY = gql`
	query SINGLE_ITEM_QUERY($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			description
			largeImage
			price
		}
	}
`;

export const SINGLE_ORDER_QUERY = gql`
	query SINGLE_ORDER_QUERY($id: ID!) {
		order(id: $id) {
			id
			total
			charge
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
			user {
				id
			}
		}
	}
`;

export const USER_ORDERS_QUERY = gql`
	query USER_ORDERS_QUERY {
		orders(orderBy: createdAt_DESC) {
			id
			total
			createdAt
			items {
				id
				title
				description
				image
				price
				quantity
			}
		}
	}
`;
