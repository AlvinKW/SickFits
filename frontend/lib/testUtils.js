import casual from 'casual';

casual.seed(777);

const mockItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'Dogs are the best!',
	description: 'Dogs',
	largeImage: 'dog.jpg',
});

const mockUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const mockOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words(),
});

const mockOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [mockOrderItem(), mockOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: mockUser(),
});

const mockCartItem = overrides => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: mockItem(),
	user: mockUser(),
	...overrides,
});

class MockLocalStorage {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

export {
	MockLocalStorage,
	mockItem,
	mockUser,
	mockCartItem,
	mockOrder,
	mockOrderItem,
};
