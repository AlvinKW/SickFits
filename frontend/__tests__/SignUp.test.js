import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';

import { mockUser } from '../lib/testUtils';
import { SIGN_UP_MUTATION } from '../lib/prismaMutations';
import { CURRENT_USER_QUERY } from '../lib/prismaQueries';
import SignUp from '../components/SignUp';

function type(wrapper, name, value) {
	wrapper.find(`input[name="${name}"]`).simulate('change', {
		target: { name, value },
	});
}

const me = mockUser();
const mocks = [
	{
		request: {
			query: SIGN_UP_MUTATION,
			variables: {
				name: me.name,
				email: me.email,
				password: 'password',
			},
		},
		result: {
			data: {
				signUp: {
					__typename: 'User',
					id: 'ABC123',
					name: me.name,
					email: me.email,
				},
			},
		},
	},
	{
		request: {
			query: CURRENT_USER_QUERY,
		},
		result: {
			data: { me },
		},
	},
];

describe('<SignUp />', () => {
	it('renders and matches the snapshot', () => {
		const wrapper = mount(
			<MockedProvider>
				<SignUp />
			</MockedProvider>
		);

		const form = wrapper.find('form');
		expect(toJSON(form)).toMatchSnapshot();
	});

	it('calls the mutation properly', async () => {
		let apolloClient = null;
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<ApolloConsumer>
					{client => {
						apolloClient = client;
						return <SignUp />;
					}}
				</ApolloConsumer>
			</MockedProvider>
		);

		await wait();
		wrapper.update();
		type(wrapper, 'name', me.name);
		type(wrapper, 'email', me.email);
		type(wrapper, 'password', 'password');

		wrapper.update();
		wrapper.find('form').simulate('submit');

		await wait();
		const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
		expect(user.data.me).toMatchObject(me);
	});
});
