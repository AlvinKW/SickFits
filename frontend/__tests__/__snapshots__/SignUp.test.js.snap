// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches the snapshot 1`] = `
<form
  className="Form-sc-11hzieo-0 ivMStP"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up For An Account
    </h2>
    <Error
      error={Object {}}
    />
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="Name"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="E-Mail"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up
    </button>
  </fieldset>
</form>
`;
